server-common/src/main/java/org/apache/kafka/server/common/DeleteRecordsPartitionStatus.java

```diff
@@ -0,0 +1,156 @@
+                                        DeleteRecordsResponseData.DeleteRecordsPartitionResult responseStatus) {
+        this.requiredOffset = requiredOffset;
+        this.responseStatus = responseStatus;
+        this.acksPending = false;
+    }
+
+    public boolean isAcksPending() {
+        return acksPending;
+    }
+
+    public void setAcksPending(boolean acksPending) {
+        this.acksPending = acksPending;
+    }
+
+
+    public DeleteRecordsResponseData.DeleteRecordsPartitionResult getResponseStatus() {
+        return responseStatus;
+    }
+
+    public long getRequiredOffset() {
+        return requiredOffset;
+    }
+
+    @Override
+    public String toString() {
+        return String.format("[acksPending: %b, error: %s, lowWatermark: %d, requiredOffset: %d]",
+                acksPending, Errors.forCode(responseStatus.errorCode()).toString(), responseStatus.lowWatermark(),
+                requiredOffset);
+    }
+
+    /**
+     * A delayed delete records operation that can be created by the replica manager and watched
+     * in the delete records operation purgatory
+     */
+    public static class DelayedDeleteRecords extends DelayedOperation {
+        private final Map<TopicPartition, DeleteRecordsPartitionStatus> deleteRecordsStatus;
+        private final BiConsumer<TopicPartition, DeleteRecordsPartitionStatus> onAcksPending;
+        private final Consumer<Map<TopicPartition, DeleteRecordsResponseData.DeleteRecordsPartitionResult>> responseCallback;
+        private static final Logger log = LoggerFactory.getLogger(DelayedDeleteRecords.class);
+
+        public DelayedDeleteRecords(long delayMs,
+                                    Map<TopicPartition, DeleteRecordsPartitionStatus> deleteRecordsStatus,
+                                    //  To maintain compatibility with dependency packages, the logic has been moved to the caller.
+                                    BiConsumer<TopicPartition, DeleteRecordsPartitionStatus> onAcksPending,
+                                    Consumer<Map<TopicPartition,
+                                            DeleteRecordsResponseData.DeleteRecordsPartitionResult>> responseCallback) {
+            super(delayMs);
+            this.onAcksPending = onAcksPending;
+            this.deleteRecordsStatus = new ConcurrentHashMap<>(deleteRecordsStatus);
+            this.responseCallback = responseCallback;
+            // first update the acks pending variable according to the error code
+            deleteRecordsStatus.forEach((topicPartition, status) -> {
+                if (status.getResponseStatus().errorCode() == Errors.NONE.code()) {
+                    // Timeout error state will be cleared when required acks are received
+                    status.setAcksPending(true);
+                    status.getResponseStatus().setErrorCode(Errors.REQUEST_TIMED_OUT.code());
+                } else {
+                    status.setAcksPending(false);
+                }
+            });
```

Comment:
We should remain this log `trace("Initial partition status for %s is %s".format(topicPartition, status))`